services/searchengine/bleveengine/search.go

```diff
@@ -0,0 +1,419 @@
+		return model.NewAppError("Bleveengine.IndexUser", "bleveengine.index_user.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	return nil
+}
+
+func (b *BleveEngine) SearchUsersInChannel(teamId, channelId string, restrictedToChannels []string, term string, options *model.UserSearchOptions) ([]string, []string, *model.AppError) {
+	if restrictedToChannels != nil && len(restrictedToChannels) == 0 {
+		return []string{}, []string{}, nil
+	}
+
+	// users in channel
+	var queries []query.Query
+	if term != "" {
+		termQ := bleve.NewPrefixQuery(strings.ToLower(term))
+		if options.AllowFullNames {
+			termQ.SetField("SuggestionsWithFullname")
+		} else {
+			termQ.SetField("SuggestionsWithoutFullname")
+		}
+		queries = append(queries, termQ)
+	}
+
+	channelIdQ := bleve.NewTermQuery(channelId)
+	channelIdQ.SetField("ChannelsIds")
+	queries = append(queries, channelIdQ)
+
+	query := bleve.NewConjunctionQuery(queries...)
+
+	uchan, err := b.UserIndex.Search(bleve.NewSearchRequest(query))
+	if err != nil {
+		return nil, nil, model.NewAppError("Bleveengine.SearchUsersInChannel", "bleveengine.search_users_in_channel.uchan.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// users not in channel
+	boolQ := bleve.NewBooleanQuery()
+
+	if term != "" {
+		termQ := bleve.NewPrefixQuery(strings.ToLower(term))
+		if options.AllowFullNames {
+			termQ.SetField("SuggestionsWithFullname")
+		} else {
+			termQ.SetField("SuggestionsWithoutFullname")
+		}
+		boolQ.AddMust(termQ)
+	}
+
+	teamIdQ := bleve.NewTermQuery(teamId)
+	teamIdQ.SetField("TeamsIds")
+	boolQ.AddMust(teamIdQ)
+
+	outsideChannelIdQ := bleve.NewTermQuery(channelId)
+	outsideChannelIdQ.SetField("ChannelsIds")
+	boolQ.AddMustNot(outsideChannelIdQ)
+
+	if len(restrictedToChannels) > 0 {
+		restrictedChannelsQ := bleve.NewDisjunctionQuery()
+		for _, channelId := range restrictedToChannels {
+			restrictedChannelQ := bleve.NewTermQuery(channelId)
+			restrictedChannelsQ.AddQuery(restrictedChannelQ)
+		}
+		boolQ.AddMust(restrictedChannelsQ)
+	}
+
+	nuchan, err := b.UserIndex.Search(bleve.NewSearchRequest(boolQ))
+	if err != nil {
+		return nil, nil, model.NewAppError("Bleveengine.SearchUsersInChannel", "bleveengine.search_users_in_channel.nuchan.error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	uchanIds := []string{}
+	for _, result := range uchan.Hits {
+		uchanIds = append(uchanIds, result.ID)
+	}
+
+	nuchanIds := []string{}
+	for _, result := range nuchan.Hits {
+		nuchanIds = append(nuchanIds, result.ID)
+	}
+
+	return uchanIds, nuchanIds, nil
+}
+
+func (b *BleveEngine) SearchUsersInTeam(teamId string, restrictedToChannels []string, term string, options *model.UserSearchOptions) ([]string, *model.AppError) {
+	if restrictedToChannels != nil && len(restrictedToChannels) == 0 {
+		return []string{}, nil
+	}
+
+	var rootQ query.Query
+	if term == "" && teamId == "" && restrictedToChannels == nil {
+		rootQ = bleve.NewMatchAllQuery()
```

Comment:
This will return all the users. Is that correct?

Reply:
Yes!